### PR TITLE
Add Reverso foundation model for zero-shot time series forecasting

### PR DESCRIPTION
**Pull request by shinfxh**
Checklist before merging this PR:
- [x] Mentioned all issues that this PR fixes or addresses.
- [x] Summarized the updates of this PR under **Summary**.
- [x] Added an entry under **Unreleased** in the [Changelog](../CHANGELOG.md).

Fixes #3034.

### Summary
- Add `ReversoModel`, a lightweight foundation model (200K-2.6M params) for zero-shot univariate time series forecasting, ported from [shinfxh/reverso](https://github.com/shinfxh/reverso)
- Pure PyTorch implementation with no external dependencies beyond what Darts already requires (replaces `flash-linear-attention` with a parallel-scan delta rule, and `FlashFFTConv` with `torch.fft`)
- Three pretrained variants on HuggingFace Hub: `reverso-nano` (200K), `reverso-small` (550K), `reverso-base` (2.6M)

## Implementation details
- `ReversoModel` extends `FoundationModel` with `HuggingFaceConnector` for weight loading
- DeltaNet attention supports both recurrent form and chunked parallel.
- Supports univariate point forecasting; multivariate targets are handled independently per component
- MIT license for ported code is included in source files

### Other Information
The CPU implementation of Reverso is still much slower than the GPU implementation at [shinfxh/reverso](https://github.com/shinfxh/reverso). For production use it is still much more recommended to use the GPU implementation!

### CODE REVIEW

**Comment by daidahao** [comment]
@shinfxh 
Thank you for the contribution and especially the chunked parallel optimisation in native PyTorch. The PR looks clean and in great shape already.

I will try to provide more comments and reviews in the coming days. In the meantime, could you update the documentation `index.rst`, `README.md`, `covariates.md` like what I did in #2980? That way, the new model would be properly advertised to our users!

File: darts/models/forecasting/reverso_model.py [hunk 1]
```diff
@@ -0,0 +1,542 @@
+"""
+Reverso
+-------
+
+Reverso is a highly parameter efficient model that achieves comparable performance with models 100x its size.
+
+A combination of long convolutions and DeltaNet sequence mixing modules are used.
```
Comment (daidahao) [inline darts/models/forecasting/reverso_model.py:7]: Header is supposed to be short. Please consider removing this line here.

Model intro should go into `ReversoModel` docstring.

File: darts/models/forecasting/reverso_model.py [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+    def __init__(
+        self,
+        seq_len: int = 2048,
+        input_token_len: int = 2048,
+        output_token_len: int = 48,
+        d_model: int = 64,
+        d_intermediate: int = 256,
+        output_bottleneck_dim: int = 48,
+        expand_v: float = 1.0,
+        state_weaving: int | bool = False,
```
Comment (daidahao) [inline darts/models/forecasting/reverso_model.py:55]: Is there a reason why `state_weaving`, `use_norm`, `learn_bias`, and `use_output_pe` could be `int` here? Can we not assume they are all bools?

File: darts/models/forecasting/reverso_model.py [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+        Returns
+        -------
+        torch.Tensor
+            Predictions of shape (batch, output_token_len, 1).
+        """
+        # min-max normalization
+        if self.use_norm:
+            x_min = x.min(1, keepdim=True)[0].detach()
+            x_max = x.max(1, keepdim=True)[0].detach()
+            x_range = torch.clamp(x_max - x_min, min=1e-5).detach()
```
Comment (daidahao) [inline darts/models/forecasting/reverso_model.py:173]: Calling `.detach()` seems unnecessary here.

File: darts/models/forecasting/reverso_model.py [hunk 4]
```diff
@@ -0,0 +1,542 @@
...
+        return out
+
+
+class ReversoModel(FoundationModel):
+    def __init__(
+        self,
+        input_chunk_length: int,
+        output_chunk_length: int,
+        output_chunk_shift: int = 0,
+        hub_model_name: str = "shinfxh/reverso-small",
```
Comment (daidahao) [inline darts/models/forecasting/reverso_model.py:262]: If `shinfxh/reverso-base` has the best performance, can we not use it as the default model? What is the trade-off here?

File: darts/models/forecasting/reverso_model.py [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+            You can perform full or partial fine-tuning of the model by setting the ``enable_finetuning`` parameter.
+            Read more in the parameter description below and in the `Fine-Tuning Examples
+            <https://unit8co.github.io/darts/examples/27-Torch-and-Foundation-Model-Fine-Tuning-examples.html>`__.
+
+        Parameters
+        ----------
+        input_chunk_length
+            Number of time steps in the past to take as a model input (per chunk). Applies to the target
+            series. For Reverso, ``input_chunk_length`` must be less than or equal to the model's context
+            length (2048 for all Reverso variants).
```
Comment (daidahao) [inline darts/models/forecasting/reverso_model.py:301]:
```suggestion
            Number of time steps in the past to take as a model input (per chunk). Applies to the target
            series. For Reverso, maximum is 2048.
```

File: darts/models/forecasting/reverso_model.py [hunk 6]
```diff
@@ -0,0 +1,542 @@
...
+        output_chunk_length
+            Number of time steps predicted at once (per chunk) by the internal model. It is not the same
+            as forecast horizon ``n`` used in ``predict()``, which is the desired number of prediction points
+            generated using either a one-shot- or autoregressive forecast. Setting ``n <= output_chunk_length``
+            prevents auto-regression.
```
Comment (daidahao) [inline darts/models/forecasting/reverso_model.py:306]:
```suggestion
        output_chunk_length
            Number of time steps predicted at once (per chunk) by the internal model. Also, the number of future values
            from future covariates to use as a model input (if the model supports future covariates). It is not the same
            as forecast horizon `n` used in `predict()`, which is the desired number of prediction points generated
            using either a one-shot- or autoregressive forecast. Setting `n <= output_chunk_length` prevents
            auto-regression.
```

File: darts/models/forecasting/reverso_model.py [hunk 7]
```diff
@@ -0,0 +1,542 @@
...
+            For Reverso, ``output_chunk_length + output_chunk_shift`` must be less than or equal to the
+            model's output token length (48 for all Reverso variants).
```
Comment (daidahao) [inline darts/models/forecasting/reverso_model.py:308]:
```suggestion
            For all variants, ``output_chunk_length + output_chunk_shift`` must be less than or equal to 48.
```

File: darts/models/forecasting/reverso_model.py [hunk 8]
```diff
@@ -0,0 +1,542 @@
...
+        output_chunk_shift
+            Optionally, the number of steps to shift the start of the output chunk into the future (relative to the
+            input chunk end). This will create a gap between the input and output. Predictions will start
+            ``output_chunk_shift`` steps after the end of the target ``series``. If ``output_chunk_shift`` is set,
+            the model cannot generate autoregressive predictions (``n > output_chunk_length``).
+        hub_model_name
+            The model ID on HuggingFace Hub. Default: ``"shinfxh/reverso-small"``.
```
Comment (daidahao) [inline darts/models/forecasting/reverso_model.py:315]: Same suggestion as above, could use the largest as default here.

File: darts/models/forecasting/reverso_model.py [hunk 9]
```diff
@@ -0,0 +1,542 @@
...
+            Other available variants: ``"shinfxh/reverso-nano"`` and ``"shinfxh/reverso-base"``.
+        hub_model_revision
```
Comment (daidahao) [inline darts/models/forecasting/reverso_model.py:317]: If the model weight is subject to changes in the future, say Reverso 1.5, I would suggest adding a commit hash here.

File: darts/models/forecasting/reverso_model.py [hunk 10]
```diff
@@ -0,0 +1,542 @@
...
+
+        return out
+
+
+class ReversoModel(FoundationModel):
+    def __init__(
+        self,
+        input_chunk_length: int,
+        output_chunk_length: int,
+        output_chunk_shift: int = 0,
```
Comment (daidahao) [inline darts/models/forecasting/reverso_model.py:261]: I would suggest adding back `likelihood` argument here, for two reasons:
- We should validate that `likelihood==None`, so the users would not accidentially set it to probabilistic. Exposing `likelihood` argument makes the validation straightforward, otherwise you need to infer from `kwargs`.
- We could highlight in the docstring the fact that Reverso is deterministic and does not support any likelihood function:
```
        likelihood
            The likelihood model to be used for probabilistic forecasts. Must be ``None`` for Reverso.
```

File: darts/models/forecasting/reverso_model.py [hunk 11]
```diff
@@ -0,0 +1,542 @@
...
+        References
+        ----------
+        .. [1] X. Fu, Y. Li, G. Papaioannou, Y. Kim. "Reverso: Efficient Time Series Foundation Models for
+                Zero-shot Forecasting", 2026. arXiv https://arxiv.org/abs/2602.17634.
+
+        Examples
+        --------
+        >>> from darts.datasets import WeatherDataset
+        >>> from darts.models import ReversoModel
+        >>> # load data in float32 format (macOS issues with float64 and PyTorch)
```
Comment (daidahao) [inline darts/models/forecasting/reverso_model.py:469]:
```suggestion
        >>> # load data in float32 format (Reverso weights are float32)
```

File: darts/models/forecasting/reverso_model.py [hunk 12]
```diff
@@ -0,0 +1,542 @@
...
+        >>> target = series['p (mbar)'][:200]
+        >>> model = ReversoModel(
+        >>>     input_chunk_length=96,
+        >>>     output_chunk_length=48,
+        >>> )
+        >>> # calling fit is still mandatory to ensure consistent number of components; however,
+        >>> # ReversoModel is training-free and the model weights are not updated
+        >>> model.fit(target)
+        >>> pred = model.predict(48)
+
```
Comment (daidahao) [inline darts/models/forecasting/reverso_model.py:481]: add print results here

File: README.md [hunk 13]
```diff
@@ -277,6 +277,7 @@ Here's a breakdown of the forecasting models currently implemented in Darts. Our
...
+| [ReversoModel](https://unit8co.github.io/darts/generated_api/darts.models.forecasting.reverso_model.html#darts.models.forecasting.reverso_model.ReversoModel)                                                                                                                                    | [Reverso paper](https://arxiv.org/abs/2602.17634), [GitHub](https://github.com/shinfxh/reverso)                                                                                                                                  | ✅ 🔴                                                          | 🔴 🔴 🔴                                                                 | 🔴 🔴                                                                      | ✅                                         |
```
Comment (daidahao) [inline README.md:280]: Reverso supports multivariate.

File: docs/source/index.rst [hunk 14]
```diff
@@ -675,6 +675,12 @@ Our regression models are designed to predict continuous numerical values, makin
      - ✅ ✅
      - ✅
      - `TimesFM 1.0 paper <https://arxiv.org/abs/2310.10688>`_, `Google blog post <https://research.google/blog/a-decoder-only-foundation-model-for-time-series-forecasting>`_
+   * - `ReversoModel <https://unit8co.github.io/darts/generated_api/darts.models.forecasting.reverso_model.html#darts.models.forecasting.reverso_model.ReversoModel>`_
+     - ✅ 🔴
```
Comment (daidahao) [inline docs/source/index.rst:679]: Reverso supports multivariate

File: darts/models/forecasting/reverso_model.py [hunk 15]
```diff
@@ -0,0 +1,542 @@
...
+            Reverso is licensed under the `MIT License <https://github.com/shinfxh/reverso/blob/main/LICENSE>`_,
+            Copyright (c) 2026 Xinghong Fu, Yanhong Li, Georgios Papaioannou, Yoon Kim.
+            By using this model, you agree to the terms and conditions of the license.
+        .. note::
+            Reverso does not support covariates natively. For multivariate time series, each component
+            is forecasted independently.
+        .. warning::
+            CPU inference is significantly slower than GPU due to the use of torch Conv instead of
+            flashfft and sequential delta-rule computation instead of fla implementation. GPU is
+            recommended for production use. See https://github.com/shinfxh/reverso/.
```
Comment (daidahao) [inline darts/models/forecasting/reverso_model.py:492]: I remembered you mentioned slowdown partly due to bfloat16. Could we recover some of slowdown by setting `pl_trainer_kwargs={"precision": "bf16-mixed" }` here?

File: darts/tests/models/forecasting/test_foundation.py [hunk 16]
```diff
@@ -456,3 +456,136 @@ def test_finetuning_all_models(self, config):
...
+    path = variant_dir / filename
+    if local_dir is None:
+        return str(path)
+    else:
+        dest_path = Path(local_dir) / filename
+        shutil.copy(path, dest_path)
+        return str(dest_path)
+
+
+class TestReversoModel:
```
Comment (daidahao) [inline darts/tests/models/forecasting/test_foundation.py:490]: All these reverso-specific tests should be in a separate file `test_reverso.py`.

File: darts/tests/models/forecasting/test_foundation.py [hunk 17]
```diff
@@ -456,3 +456,136 @@ def test_finetuning_all_models(self, config):
...
         assert preds.shape == (6, self.series.n_components * len(quantiles), 1)
+
+
+reverso_artefacts_dir = (Path(__file__).parent / "artefacts" / "reverso").absolute()
+
+reverso_variant_dirs = {
+    "shinfxh/reverso-nano": reverso_artefacts_dir / "tiny_reverso_nano",
+    "shinfxh/reverso-small": reverso_artefacts_dir / "tiny_reverso_small",
+    "shinfxh/reverso-base": reverso_artefacts_dir / "tiny_reverso_full",
+}
```
Comment (daidahao) [inline darts/tests/models/forecasting/test_foundation.py:467]: Do you need three mock models here? I think one "tiny" reverso-nano would be enough.

File: darts/tests/models/forecasting/test_foundation.py [hunk 18]
```diff
@@ -456,3 +456,136 @@ def test_finetuning_all_models(self, config):
...
+        return str(path)
+    else:
+        dest_path = Path(local_dir) / filename
+        shutil.copy(path, dest_path)
+        return str(dest_path)
+
+
+class TestReversoModel:
+    series = generate_series(n_variables=2, length=100, prefix="A")
+
```
Comment (daidahao) [inline darts/tests/models/forecasting/test_foundation.py:492]: I think the current test suite might be incomplete. Looking at `TestTimesFM2p5Model`, the following tests seem to be missing here:
- `test_fidelity`: this is important to make sure the implementaton is correct here, against the original. I recommend using the actual reverso-full for this one.
- `test_default`: default model should be deterministic.
- `test_probabilistic`: reverso cannot be probabilistic, so an error is expected here.
- `test_multivariate`: reverso shoould support multivariate.
- `test_covariates`: reverso does not support covariates, so errors are expected.
- `test_multiple_series`: reverso supports forecasting on multiple time series at the same time.

Except for `test_fidelity`, all other tests should use a tiny mock reverso to reduce overhead.

File: CHANGELOG.md [hunk 19]
```diff
@@ -9,6 +9,7 @@ but cannot always guarantee backwards compatibility. Changes that may **break co
 
 ### For users of the library:
 
+- Added `ReversoModel`, a new foundation model for zero-shot time series forecasting. Reverso is a highly parameter-efficient model (200K-2.6M params) that matches accuracy of models 100x its size. [#3034](https://github.com/unit8co/darts/issues/3034) by [Xinghong Fu](https://github.com/shinfxh).
```
Comment (daidahao) [inline CHANGELOG.md:12]:
```suggestion
- Added new forecasting model `ReversoModel`: a family of highly parameter-efficient (200K-2.6M) foundation models that matches accuracy of models 100x their sizes. It supports univariate, multivariate, and multiple time series forecasting without training and can be fine-tuned on your own data. [#3034](https://github.com/unit8co/darts/issues/3034) by [Xinghong Fu](https://github.com/shinfxh).
```

File: darts/models/forecasting/reverso_model.py [hunk 20]
```diff
@@ -0,0 +1,542 @@
...
+        """
+        # B: batch size, L: input chunk length, C: target components
+        x_past, _, _ = x_in
+        B, L, C = x_past.shape
+
+        # channel independence: (B, L, C) -> (B*C, L)
+        x = x_past.permute(0, 2, 1).reshape(-1, L)
+
+        # left-pad with per-series first value to seq_len
+        if L < self.seq_len:
```
Comment (daidahao) [inline darts/models/forecasting/reverso_model.py:234]: Just curious, do we have to pad the context here? I imagine removing the padding could speed up inference a lot for small `input_chunk_length`. But we might introduce train-test mismatch here if the model only performs well with 2048 input.

**Review by daidahao — CHANGES_REQUESTED** [review]
Thank you @shinfxh for this PR and congratulations on releasing the models.

The PR is in very good shape already so I mostly made minor comments on wordings. The main changes I am requesting would be the unit tests that cover fidelity tests, invalid usage, and fine-tuning, etc. Also, you might need to merge with main branch, to use the lazy import for `ReversoModel`.

Because Reverso is branded as efficient, I would suggest adding some tips to further boost its inference speed, such as using GPU/MPS, bfloat16, which can be set via `pl_trainer_kwargs`, and using reverso-nano via `hub_model_name`.

**Comment by daidahao** [comment]
@shinfxh Let me know if you need any assistance.

**Comment by shinfxh** [comment]
Thanks for the comments, will get to it soon, thanks for the wait!

**Comment by daidahao** [comment]
@shinfxh Long time! Can I get an update on this?

**Comment by daidahao** [comment]
Hi @shinfxh, we have not received updates from you for a while. Would you still like to proceed with this PR? If so, do you need any assistance in terms of coding or review from me?
